src/engine/GroupByImpl.cpp

```diff
@@ -1698,49 +1711,47 @@ IdTable GroupByImpl::createResultFromHashMap(
 // Visitor function to extract values from the result of an evaluation of
 // the child expression of an aggregate, and subsequently processing the
 // values by calling the `addValue` function of the corresponding aggregate.
-static constexpr auto makeProcessGroupsVisitor =
-    [](size_t blockSize,
-       const sparqlExpression::EvaluationContext* evaluationContext,
-       const GroupByImpl::GroupLookupResult& groupLookupResult) {
-      return CPP_template_lambda(blockSize, evaluationContext,
-                                 &groupLookupResult)(typename T, typename A)(
-          T && singleResult, A & aggregationDataVector)(
-          requires sparqlExpression::SingleExpressionResult<T> &&
-          VectorOfAggregationData<A>) {
-        // Create pairs of [value, rowIndex] from generator
-        auto generator = sparqlExpression::detail::makeGenerator(
-            std::forward<T>(singleResult), blockSize, evaluationContext);
-
-        const auto& matches = groupLookupResult.matchedRowsToGroups_;
-        // Ensure matches are sorted by rowIndex, as we will iterate
-        // through the generator in ascending order of rowIndex
-        AD_EXPENSIVE_CHECK(ql::ranges::is_sorted(
-            matches.begin(), matches.end(), {},
-            &GroupByImpl::GroupLookupResult::RowToGroup::rowIndex_));
-
-        // Process each matching row
-        // TODO<Benke> Potentially a waste of resources!
-        //   The code evaluates the sub-expressions for the non-matching rows
-        //   twice, because here they are evaluated then discarded, and later on
-        //   they are evaluated again from the materialized and sorted result.
-        auto matchIt = matches.begin();
-        size_t currentRow = 0;
-        for (const auto& value : generator) {
-          if (matchIt != matches.end() && currentRow == matchIt->rowIndex_) {
-            aggregationDataVector.at(matchIt->groupIndex_)
-                .addValue(value, evaluationContext);
-            ++matchIt;
-          }
-          ++currentRow;
-        }
-      };
-    };
+constexpr auto GroupByImpl::makeProcessGroupsVisitor(
+    size_t blockSize,
+    const sparqlExpression::EvaluationContext* evaluationContext,
+    const GroupLookupResult& groupLookupResult) {
+  return CPP_template_lambda(blockSize, evaluationContext, &groupLookupResult)(
+      typename T, typename A)(T && singleResult, A & aggregationDataVector)(
+      requires sparqlExpression::SingleExpressionResult<T> &&
+      VectorOfAggregationData<A>) {
+    const auto& matchedRows = groupLookupResult.matchedRows_;
+    const auto& groupIndexes = groupLookupResult.groupIndexes_;
```

Comment:
Ah, you need the matched rows separately for your new sparse generator thing. you can keep it for now, as the more efficient fix for this also requires to have them separately probably.